Change default font and fix potential buffer overflow

For X11 multi-byte we want to ensure there is a good default font selected,
so we change the default font to font variants present on openSUSE systems.

Prevent buffer overflow when changing the default font to a maliciously
crafted font name.

Make output for missing character sets less verbose by preferring debug
prints
---
 src/gplt_x11.c |   33 +++++++++++++++++++++++++++------
 1 file changed, 27 insertions(+), 6 deletions(-)

Index: src/gplt_x11.c
===================================================================
--- src/gplt_x11.c.orig
+++ src/gplt_x11.c
@@ -106,15 +106,16 @@
 #include "stdfn.h"
 #include "gp_types.h"
 
+#ifndef _GNU_SOURCE
+#define _GNU_SOURCE
+#endif
+#include <string.h>
 #include <X11/Xos.h>
 #include <X11/Xlib.h>
 #include <X11/Xresource.h>
 #include <X11/Xutil.h>
 #include <X11/Xatom.h>
 #include <X11/keysym.h>
-#ifdef USE_X11_MULTIBYTE
-# include <X11/Xlocale.h>
-#endif
 #include <X11/XKBlib.h>	/* for XkbKeycodeToKeysym */
 
 #include <assert.h>
@@ -122,6 +123,12 @@
 #include "gplt_x11.h"
 #include "version.h"
 
+#ifdef USE_X11_MULTIBYTE
+# include <locale.h>
+# include <langinfo.h>
+# include <X11/Xlocale.h>
+#endif
+
 #ifdef EXPORT_SELECTION
 # undef EXPORT_SELECTION
 #endif /* EXPORT SELECTION */
@@ -526,8 +533,8 @@ static char *gpFallbackFont(void);
 static int gpXGetFontascent(XFontStruct *cfont);
 
 enum set_encoding_id encoding = S_ENC_DEFAULT; /* EAM - mirrored from core code by 'QE' */
-static char default_font[196] = { '\0' };
-static char default_encoding[16] = { '\0' };
+static char default_font[256] = { '\0' };
+static char default_encoding[64] = { '\0' };
 
 #define Nwidths 10
 static unsigned int widths[Nwidths] = { 1, 0, 0, 0, 0, 0, 0, 0, 0, 0 };
@@ -2294,6 +2301,8 @@ exec_cmd(plot_struct *plot, char *comman
 		/* Save the request default font */
 		c = &(buffer[strlen(buffer)-1]);
 		while (*c <= ' ') *c-- = '\0';
+		if (strlen(&buffer[2]) == 0)
+		    break;
 		strncpy(default_font, &buffer[2], sizeof(default_font)-1);
 		FPRINTF((stderr, "gnuplot_x11: exec_cmd() set default_font to \"%s\"\n", default_font));
 		break;
@@ -5814,7 +5823,7 @@ XFontStruct *gpXLoadQueryFont (Display *
 	    }
 #else
 	    if (first_time) {
-		fprintf(stderr,"gnuplot_x11: Some character sets not available\n");
+		FPRINTF(("gnuplot_x11: Some character sets not available\n"));
 		first_time = FALSE;
 	    }
 	    while (n_miss-- > 0)
@@ -5966,6 +5975,18 @@ pr_font( char *fontname )
     }
 
 #ifdef USE_X11_MULTIBYTE
+    if ((!fontname || !(*fontname)) && multibyte_fonts_usable) {
+	const char * codeset = nl_langinfo(CODESET);
+	if (strcasestr(codeset, "UTF-8")) {
+		const char* utf8 = "mbfont:-*-sazanami*mincho-medium-r-normal--16-* \
+-*-mincho-medium-r-normal--16-*,-*-verdana-medium-r-normal--16-* \
+-*-dejavu*sans-medium-r-normal--16-*,-*-medium-r-normal--16-*";
+		strncpy(default_font, utf8, strlen(utf8)+1);
+	}
+    }
+#endif
+
+#ifdef USE_X11_MULTIBYTE
     if (fontname && strncmp(fontname, "mbfont:", 7) == 0) {
 	if (multibyte_fonts_usable) {
 	    usemultibyte = 1;
